src/osd/ECCommon.cc

```diff
@@ -539,11 +574,19 @@ void ECCommon::ReadPipeline::objects_read_and_reconstruct(
   }
 
   map<hobject_t, set<int>> obj_want_to_read;
-  set<int> want_to_read;
-  get_want_to_read_shards(&want_to_read);
     
   map<hobject_t, read_request_t> for_read_op;
   for (auto &&to_read: reads) {
+    set<int> want_to_read;
+    if (cct->_conf->osd_ec_partial_reads) {
+      for (const auto& single_region : to_read.second) {
+        get_min_want_to_read_shards(single_region.offset,
+				    single_region.size,
+				    &want_to_read);
+      }
+    } else {
+      get_want_to_read_shards(&want_to_read);
+    }
     map<pg_shard_t, vector<pair<int, int>>> shards;
     int r = get_min_avail_to_read_shards(
```

Comment:
I don't quite understand how this new logic interacts with the case where we're missing one of the data chunks.  Suppose the PG is a 4+2 code with a chunk/stripe size of 1k/4k and the acting set contains shards 0, 1, 2, 4, 5 (it's missing the last data chunk 3).  We're serving a read on 3k~1k.

That should be exactly the 4th chunk, so get_min_want_to_read_shards should populate want_to_read with {3}.  However, 3 isn't available.  Consequently, get_min_avail_to_read_shards here is going to generate something like {0, 1, 2, 4}.

Is it just that ECCommon::ReadPipeline::do_read_op will end up calling chunk_aligned_offset_len_to_chunk for all of them?  I guess that works out even with chunk aligned offsets?

Reply:
ECUtil::decode will still get the original want_to_read, so it'll go ahead and produce 3 and then fulfill the read.  I think that works out?

Reply:
Is it the case that this implementation will always request the same boundaries from every OSD?  That is, suppose we're serving a read on the above pg 1k~7k.  Shards 2, 3, and 4 need chunks 0 and 1, but shard 0 only needs chunk 1.  I think this implementation will end up request 0 and 1 from all 4 though, right?  That's not necessarily a problem given how simple this PR ended up, just confirming my understanding.

Reply:
> Is it just that ECCommon::ReadPipeline::do_read_op will end up calling chunk_aligned_offset_len_to_chunk for all of them? I guess that works out even with chunk aligned offsets?

Yes, that's my understanding. The responsible part of the logic is:

```cpp
    for (const auto& read : i->second.to_read) {
      auto p = make_pair(read.offset, read.size);
      pair<uint64_t, uint64_t> chunk_off_len = sinfo.chunk_aligned_offset_len_to_chunk(p);
      for (auto k = i->second.need.begin();
           k != i->second.need.end();
           ++k) {
        messages[k->first].to_read[i->first].push_back(
          boost::make_tuple(
            chunk_off_len.first,
            chunk_off_len.second,
            read.flags));
      }
      ceph_assert(!need_attrs);
    }
```

`k` is shard id, `i` is `std::pair<hobject_t, read_request_t>`; `chunk_off_len` doesn't depend on particular `k` – it's calculated once for all of them.

> ECUtil::decode will still get the original want_to_read, so it'll go ahead and produce 3 and then fulfill the read. I think that works out?

That's the intention behind https://github.com/ceph/ceph/pull/55196/commits/09a88cbff4180bd7cd7040ec6cc717959d742e67.

> Is it the case that this implementation will always request the same boundaries from every OSD? 

I think so.